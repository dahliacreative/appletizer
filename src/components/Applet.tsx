import React, { useEffect } from "react";
import { IApplet } from "../methods/initializeApplet";

export interface IAppletProps extends IApplet {
  host: string;
  name: string;
}

export const Applet: React.FC<IAppletProps> = ({
  name,
  host,
  history,
  context,
}) => {
  const APPLICATION_ID = `applet-${name}`;
  const SCRIPT_ID = `applet-script-${name}`;
  const MANIFEST = `${host}/asset-manifest.json`;
  const MOUNT = `mount${name}`;
  const UNMOUNT = `unmount${name}`;
  const KEY = `applet-${name}-${Date.now()}`;

  useEffect(() => {
    const renderApplet = () => {
      (window as any)[MOUNT](APPLICATION_ID, history, context);
    };

    if (document.getElementById(SCRIPT_ID)) {
      renderApplet();
      return;
    }

    fetch(MANIFEST)
      .then((res) => res.json())
      .then((manifest) => {
        const script = document.createElement("script");
        script.id = SCRIPT_ID;
        script.crossOrigin = "";
        script.src = `${host}${manifest.files["main.js"]}`;
        script.onload = renderApplet;
        document.body.appendChild(script);
      });

    return () => {
      // Ignoring this line because it's not possible to test
      // due to a bug in testing-library where effect cleanups
      // are not called when unmounting.
      // istanbul ignore next
      (window as any)[UNMOUNT] && (window as any)[UNMOUNT]();
    };
  });

  return <main id={APPLICATION_ID} key={KEY} />;
};
