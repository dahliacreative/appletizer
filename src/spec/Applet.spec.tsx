import React from "react";
import fetchMock from "jest-fetch-mock";
import { render } from "@testing-library/react";
import { Applet } from "../components/Applet";
import { createBrowserHistory } from "history";
import { wait } from "@testing-library/user-event/dist/utils";

fetchMock.enableMocks();

describe("Applet", () => {
  it("should fetch the manifest", () => {
    (fetch as jest.Mock).mockResolvedValue({
      json: () => ({ files: { "main.js": "main.js" } }),
    });

    const { container } = render(
      <Applet
        name="Test"
        host="http://localhost:3001"
        history={createBrowserHistory()}
      />
    );

    expect(container).toMatchSnapshot();
  });

  it("mount the application when the script is already in the page", () => {
    const script = document.createElement("script");
    script.id = "applet-script-Test";
    document.head.appendChild(script);

    const mockMount = jest.fn();
    (window as any).mountTest = mockMount;

    render(
      <Applet
        name="Test"
        host="http://localhost:3001"
        history={createBrowserHistory()}
        context={{}}
      />
    );

    expect(mockMount).toHaveBeenCalledWith(
      "applet-Test",
      expect.any(Object),
      {}
    );
  });
});
