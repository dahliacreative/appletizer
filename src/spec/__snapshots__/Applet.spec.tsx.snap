// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Applet should fetch the manifest 1`] = `
<div>
  <main
    id="applet-Test"
  />
</div>
`;
